_stbt/core.py

```diff
@@ -1162,24 +1163,22 @@ def wait_until(callable_, timeout_secs=10, interval_secs=0, stable_secs=0):
     :type interval_secs: int or float, in seconds
     :param interval_secs: Delay between successive invocations of ``callable_``.
 
+    :param predicate: A function that takes a single value. It will be given
+        the return value from ``callable_``. The return value of *this* function
+        will then be used to determine truthiness. ``wait_until`` will still
+        return the original value from ``callable_``, not the predicate value.
+
     :type stable_secs: int or float, in seconds
     :param stable_secs: Wait for ``callable_``'s return value to remain the same
-        (as determined by ``==``) for this duration before returning. This can
-        be used to wait for a `FrameObject` to stabilise.
+        (as determined by ``==``) for this duration before returning. If
+        ``predicate`` is also given, the values returned from ``predicate``
+        will be compared.
 
     :returns: The return value from ``callable_`` (which will be truthy if it
-        succeeded, or falsey if ``wait_until`` timed out).
-
-        If ``stable_secs`` is specified then ``wait_until`` returns:
-
-        * ``callable_``'s return value if it was truthy and stable. This will
-          be the *first* stable value, so that you can use ``wait_until`` for
-          performance measurements (for example to measure the time for an
-          animation to complete).
-        * ``None`` if ``callable_``'s return value was truthy but not stable
-          before we reached ``timeout_secs``.
-        * ``callable_``'s last return value if it was falsey when we reached
-          ``timeout_secs``.
+        succeeded, or falsey if ``wait_until`` timed out). If the value was
+        truthy when the timeout was reached but it failed the ``predicate`` or
+        ``stable_secs`` conditions (if any) then ``wait_until`` returns
+        ``None``.
```

Comment:
So the motivation for returning `None` is for use cases like this:

```
assert wait_until(Guide, predicate=lambda x: x.selection)
```

but it would prevent use cases like:

```
mr = wait_until(is_motion, predicate=lambda x: not x)
assert not mr, "Motion found in region %r" % mr.region
```

Seems reasonable.  I guess that our `stable_secs` policy of returning the first truthy result is fundamentally incompatible with returning a truthy value when the predicate fails.  I suppose that predicate just cannot be used when waiting for `not` something because of the difficulties distinguishing whether a timeout happened or not.